Removing the make_module script.

z-a-f opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #24151 Replacing axis with dim in quantized cat
* **#23635 Removing the make_module script.**

n128675

Differential Revision: [D16593364](https://our.internmc.facebook.com/intern/diff/D16593364)

Commit:
Removing the make_module script.

It appears it is the same complexity to add new modules using a base class and using a generation script.

Commit:
Update on "Removing the make_module script."


It appears it is the same complexity to add new modules using a base class and using a generation script.

Differential Revision: [D16593364](https://our.internmc.facebook.com/intern/diff/D16593364)

Commit:
Update on "Removing the make_module script."


It appears it is the same complexity to add new modules using a base class and using a generation script.

Differential Revision: [D16593364](https://our.internmc.facebook.com/intern/diff/D16593364)

jerryzh168 commented on torch/nn/modules/_wrapped_modules.py:
```diff
@@ -0,0 +1,21 @@
+class _BaseWrapperModule(Module):
+    def __init__(self):
+        super(_BaseWrapperModule, self).__init__()
+
```
should we do torch.operation = None or something

jerryzh168 commented on torch/nn/modules/_wrapped_modules.py:
```diff
@@ -0,0 +1,21 @@
+        super(_BaseWrapperModule, self).__init__()
+
+    def forward(self, *args):
+        return torch.operation(*args)
```
self.operation?

Commit:
Update on "Removing the make_module script."


It appears it is the same complexity to add new modules using a base class and using a generation script.

Differential Revision: [D16593364](https://our.internmc.facebook.com/intern/diff/D16593364)

Commit:
Update on "Removing the make_module script."


It appears it is the same complexity to add new modules using a base class and using a generation script.

Differential Revision: [D16593364](https://our.internmc.facebook.com/intern/diff/D16593364)

Commit:
Update on "Removing the make_module script."


It appears it is the same complexity to add new modules using a base class and using a generation script.

Differential Revision: [D16593364](https://our.internmc.facebook.com/intern/diff/D16593364)

jerryzh168 commented on torch/nn/quantized/modules/_wrapped_modules.py:
```diff
@@ -0,0 +1,46 @@
+                              zero_point=self.zero_point)
+
+    @classmethod
+    def from_float(cls, mod):
```
e.g. https://github.com/pytorch/pytorch/blob/master/torch/nn/qat/modules/conv.py#L40

zafartahirov replied:
If qconfig is provided, should the scale and zero_point come from `qconfig`?

jerryzh168 replied:
yes

jerryzh168 replied:
actually, this doesn't make sense for activation, what we want to pass in is not qconfig, but a 'quantizer' which doesn't exist yet. we can do this later.

jerryzh168 commented on torch/nn/quantized/modules/_wrapped_modules.py:
```diff
@@ -0,0 +1,46 @@
+
+r"""Add module wraps torch.ops.quantized.add."""
+class Add(_BaseWrapperModule):
+    __FLOAT_MODULE = torch.nn.modules.Add
```
nit: torch.nn.Add?

jerryzh168 commented on torch/nn/quantized/modules/_wrapped_modules.py:
```diff
@@ -0,0 +1,46 @@
+from __future__ import unicode_literals
+
+import torch
+from torch.nn.modules import Module
```
nit: torch.nn import Module?

zafartahirov replied:
I think, that potentially can cause circular deps -- lemme check if I can

dzhulgakov commented on torch/nn/modules/_wrapped_modules.py:
```diff
@@ -0,0 +1,23 @@
+from torch.nn.modules import Module
+
+r"""Base class for all wrapper modules."""
+class _BaseWrapperModule(Module):
```
this class is likely not going to be scriptable - plesae double-check

jerryzh168 replied:
actually I think we probably don't need them to be scriptable since these are not used in graph mode, or are you referring to the case when we quantize in eager mode and then script it?

dzhulgakov replied:
Yes, unless there's a good reason not to, our default should be that everything is scriptable.

In this case, it's a very likely case that I want to quantize using eager mode, tweak my model and script the result

dzhulgakov reviewed (changes requested):
Unittests that the class is scriptable? (the whole point of getting rid of magic)

soumith commented on torch/nn/modules/__init__.py:
```diff
@@ -53,10 +55,7 @@
-    'Flatten'
+    'Flatten',
+    # Wrapped modules
+    'Add'
```
why are you adding an `nn.Add`?

dzhulgakov commented:
Btw, this PR could have benefited from:
- better description that provides a bit of context (e.g. copy-paste from quantization design doc)
- unittest to verify that the module works and is scriptable (at least one of them)

Commit:
Update on "Removing the make_module script."


It appears it is the same complexity to add new modules using a base class and using a generation script.

Differential Revision: [D16593364](https://our.internmc.facebook.com/intern/diff/D16593364)

dzhulgakov commented on torch/nn/quantized/modules/wrapper_module.py:
```diff
@@ -0,0 +1,48 @@
+        self.operation = operation
+
+    def forward(self, x):
+        return self.operation(*x)
```
cool, didn't know our JIT is smart enough to resolve `self.operation` (original reason to ask for explicit module)

why are you unpacking the args though? it's confusing to user and doesn't work with JIT (as it doesn't support varargs).

I guess we can just hard-code 2 arguments then or have per-num-arguments versions.

zafartahirov replied:
I am not 100% sure if we will support unary ops through this mechanism. This is more of a "generic" implementation.

zafartahirov replied:
I think it works with variable length, but only when using `torch.jit.trace`

dzhulgakov replied:
I'd rather make it work with Script too. You can create a two-argument version for now and hide it under `make_wrapper` function. In future, it'd be possible to extend to more arguments if needed (by inspecting schema of the passed operation)

zafartahirov replied:
We don't have the `make_wrapper` anymore -- it's only a wrapper class :/

dzhulgakov replied:
again, I think having them scriptable is important - and starting with two-argument ops is fine. You could introduce `make_wrapper` to allow future insertion point for other number of arguments

dzhulgakov reviewed (changes requested):
if you claim it's scriptable - where's the unittest to verify it's so?

Commit:
Update on "Removing the make_module script."


It appears it is the same complexity to add new modules using a base class and using a generation script.

Differential Revision: [D16593364](https://our.internmc.facebook.com/intern/diff/D16593364)

Commit:
Update on "Removing the make_module script."


It appears it is the same complexity to add new modules using a base class and using a generation script.

Differential Revision: [D16593364](https://our.internmc.facebook.com/intern/diff/D16593364)

Commit:
Update on "Removing the make_module script."


It appears it is the same complexity to add new modules using a base class and using a generation script.

Differential Revision: [D16593364](https://our.internmc.facebook.com/intern/diff/D16593364)

Commit:
Update on "Removing the make_module script."


It appears it is the same complexity to add new modules using a base class and using a generation script.

Differential Revision: [D16593364](https://our.internmc.facebook.com/intern/diff/D16593364)

Commit:
Update on "Removing the make_module script."


It appears it is the same complexity to add new modules using a base class and using a generation script.

Differential Revision: [D16593364](https://our.internmc.facebook.com/intern/diff/D16593364)

Commit:
Update on "Removing the make_module script."


It appears it is the same complexity to add new modules using a base class and using a generation script.

Differential Revision: [D16593364](https://our.internmc.facebook.com/intern/diff/D16593364)

raghuramank100 commented on torch/nn/quantized/modules/wrapper_module.py:
```diff
@@ -0,0 +1,70 @@
+        self.operation = operation
+
+    def forward(self, x):
+        return self.operation(x)
```
Need to have type annotation for this to work
```
class UnaryWrapper(nn.Module):
    def __init__(self, operation):
        super(UnaryWrapper, self).__init__()
        self.operation = operation

    def forward(self, x):
        # type: (List[Tensor]) -> Tensor
        return self.operation(x)
```

dzhulgakov replied:
This annotation is only valid for Cat (which takes the list). Other unary ops, shall we choose to quantize them, should have `Tensor` annotation.

I really think it's should be an implementation details. We can expose a single function `make_wrapper(torch.something)` and have internal dictionary dispatch to pick the right wrapper (out of a few - like binary, unary, list). All these wrappers aren't that generic anyway since _FLOAT_MODULES explicitly lists supported functions.

raghuramank100 commented on torch/nn/quantized/modules/wrapper_module.py:
```diff
@@ -0,0 +1,70 @@
+        self.register_buffer('scale', torch.tensor([1.0], dtype=torch.double))
+        self.register_buffer('zero_point', torch.tensor([0], dtype=torch.long))
+
+    def forward(self, x):
```
Need type annotation

raghuramank100 commented on test/test_quantization.py:
```diff
@@ -290,6 +290,72 @@ def checkQuantized(model):
+
+        BinaryWrapper = nnq.modules.wrapper_module.BinaryWrapper
+        QBinaryWrapper = nnq.modules.wrapper_module.QuantizedBinaryWrapper
+        script_module = torch.jit.trace(BinaryWrapper(torch.add),
```
Need to have check_trace=False for tracing to work. I have filed a github issue () to track this bug.
https://github.com/pytorch/pytorch/issues/23993

raghuramank100 commented on test/test_quantization.py:
```diff
@@ -290,6 +290,72 @@ def checkQuantized(model):
+        QUnaryWrapper = nnq.modules.wrapper_module.QuantizedUnaryWrapper
+        script_module = torch.jit.trace(UnaryWrapper(torch.cat),
+                                        (torch.tensor(0), torch.tensor(0)))
+        script_qmodule = torch.jit.trace(QUnaryWrapper(torch.cat),
```
These tests will not work for tracing or scripting. This is because the operators can take dynamic length inputs, one can trace with a tuple of length two and then run with a tuple of different length. A test case that corresponds to actual use is:

An notebook with the outputs is at: https://our.internmc.facebook.com/intern/anp/view/?id=128232
```python
import torch.nn as nn

class UnaryWrapper(nn.Module):
    def __init__(self, operation):
        super(UnaryWrapper, self).__init__()
        self.operation = operation

    def forward(self, x):
        # type: (List[Tensor]) -> Tensor
        return self.operation(x)


class BinaryWrapper(nn.Module):
    def __init__(self, operation):
        super(BinaryWrapper, self).__init__()
        self.operation = operation

    def forward(self, x,y):
        return self.operation(x,y)


class testModel(nn.Module):
    def __init__(self):
        super(testModel, self).__init__()
        self.mycat = UnaryWrapper(torch.cat)
        self.myadd= BinaryWrapper(torch.add)
        self.mymul = BinaryWrapper(torch.mul)
    
    def forward(self,x):
        y = self.mycat([x ,x ,x])
        z = self.myadd(y, y)
        z = self.mymul(z,z)
        return z
  ```

raghuramank100 reviewed (changes requested):
Based on discussions with @jamesr66a have suggested changes that are required for tracing and scripting to work.

Commit:
Update on "Removing the make_module script."


It appears it is the same complexity to add new modules using a base class and using a generation script.

Differential Revision: [D16593364](https://our.internmc.facebook.com/intern/diff/D16593364)

Commit:
Update on "Removing the make_module script."


It appears it is the same complexity to add new modules using a base class and using a generation script.

Differential Revision: [D16593364](https://our.internmc.facebook.com/intern/diff/D16593364)

Commit:
Update on "Removing the make_module script."


n128675

Differential Revision: [D16593364](https://our.internmc.facebook.com/intern/diff/D16593364)

Commit:
Update on "Removing the make_module script."


n128675

Differential Revision: [D16593364](https://our.internmc.facebook.com/intern/diff/D16593364)

raghuramank100 commented on test/test_quantization.py:
```diff
@@ -290,6 +292,34 @@ def checkQuantized(model):
+class ScriptabilityTest(QuantizationTestCase):
+    def test_wrapper_traceable(self):
+        model_under_test = ModForWrapping()
+        x = torch.rand(10)
```
Can you combine test_wrapper_traceable and test_wrapper_scriptable?

raghuramank100 commented on torch/quantization/quantize.py:
```diff
@@ -70,7 +70,10 @@ def add_observer(module, skip_list=DEFAULT_SKIP_LIST):
     """
     for child in module.children():
-        add_observer(child)
+        if type(child) == nnq.FloatFunctional:
```
This code should be called once you know you are at a leaf module and can check for qconfig:
Remove this change here and insert below:
```python
 # Insert observers only for leaf nodes, note that this observer is for
    # the output of the module, for input QuantStub will observe them
    if hasattr(module, 'qconfig') and module.qconfig is not None and \
       len(module._modules) == 0 and type(module) not in skip_list:
        # observer and hook will be gone after we swap the module
        if type(module) == nnq.FloatFunctional:
              module.observer = module.qconfig.activation()
      else:
           module.add_module('observer', module.qconfig.activation())
           module.register_forward_hook(_observer_forward_hook)
```

zafartahirov replied:
This is already there later in the code. At the same time, FloatFunctional is not expected to have children, so it will always be a child. The reason is that the FloatFunctional will raise an error if `forward` is called, which means it's always a leaf. At the same time, because this this is a recursive function, semantically it is the same

zafartahirov replied:
Made the change anyway :)

Commit:
Update on "Removing the make_module script."


n128675

Differential Revision: [D16593364](https://our.internmc.facebook.com/intern/diff/D16593364)

raghuramank100 reviewed (approved):
Approved, please address comments before landing.

Commit:
Update on "Removing the make_module script."


n128675

Differential Revision: [D16593364](https://our.internmc.facebook.com/intern/diff/D16593364)

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/qconcat.cpp:
```diff
@@ -22,7 +22,7 @@ bool is_valid_quantization_scheme(Tensor t) {
 Tensor quantized_cat(
     const std::vector<Tensor>& qxs,
-    int64_t axis,
+    int64_t dim,
```
it's probably fine now, but in general - don't mix unrelated changes in the same PR

zafartahirov replied:
My bad -- this was bothering me a lot, and I was too lazy to have a stack :) Will not do that again!

dzhulgakov commented on torch/nn/quantized/modules/functional_modules.py:
```diff
@@ -0,0 +1,76 @@
+        assert self.op_name == 'add', \
+            "Running add while initialized as {}".format(self.op_name)
+        r = torch.add(x, y)
+        self.observer(r)
```
this might not work with QAT, because I think observer returns a value override in that case - cc @jerryzh168 . Maybe put a TODO now and address it in a separate diff?

raghuramank100 replied:
Good catch!, we will need to return a value for QAT.  @jerryzh168 : Even the existing code has a problem: We are calling fake quant in the forward hook and this needs to change numerics, which is not allowed.

zafartahirov replied:
So should that be `r = self.observer(r)`?

dzhulgakov replied:
QAT observers return overrides

zafartahirov replied:
Does that mean that QAT and PTQ observers have different semantics?

dzhulgakov replied:
That's the current observer semantics, if non-None is returned, it overrides the return: https://github.com/pytorch/pytorch/blob/77c08aa46c3f3460b95b89cbe357b99180bc824d/torch/nn/modules/module.py#L549-L551

zafartahirov replied:
Got it! I will add a TODO, and can send a diff later on, while bringing @jerry39213gh into the loop

raghuramank100 replied:
This should be a very simple diff: https://github.com/pytorch/pytorch/pull/24194. I just sent it out. Also, it should be @jerryzh168

dzhulgakov commented on torch/nn/quantized/modules/functional_modules.py:
```diff
@@ -0,0 +1,76 @@
+        # type: (List[Tensor], Optional[int]) -> Tensor
+        assert self.op_name == 'cat', \
+            "Running cat while initialized as {}".format(self.op_name)
+        if dim is None:
```
add a TODO to fix op registration. This logic here feels out of place

zafartahirov replied:
The registtration will support optional arguments. The thing is that I still believe we should resolve it here, instead of doing it  inside of the kernel. The reason is that Once the documentation is parsed, the default values will be shown. But if we do it in the kernel, the default value will be hidden from the user.

zafartahirov replied:
I can make it optional with default value of 0

dzhulgakov replied:
If it's a bug - then file an issue against JIT/op bindings to fix it :) It feels out of place here.

zafartahirov replied:
I don't think it's a bug, as it is documented in the [Type Refinement](https://pytorch.org/docs/stable/jit.html#optional-type-refinement) section of the jit docs. @jamesr66a Is there a way to use optional args, while dropping the `if None` test?

dzhulgakov replied:
Actually, you're right, it's not a optional arg, it's an arg with default value: https://github.com/pytorch/pytorch/blob/77c08aa46c3f3460b95b89cbe357b99180bc824d/aten/src/ATen/native/native_functions.yaml#L399

Then why you just don't duplicate the signature to be the same, i.e. `int dim = 0`?

jamesr66a replied:
I think the ideal solution in python would be to use kwargs and forward them directly:

```
import torch

dim = None

x = [torch.rand(3, 4) for i in range(3)]

def cat(x, **kwargs):
    torch.cat(x, **kwargs)

cat(x, dim=1)
cat(x)
```

However, this isn't supported by script, so what's used here is fine. @zafartahirov also proposed making the default `dim` equal to ` in the function signature here, which would probably also work

dzhulgakov commented on torch/nn/quantized/modules/functional_modules.py:
```diff
@@ -0,0 +1,76 @@
+    'cat'
+}
+
+class FloatFunctional(torch.nn.Module):
```
comments on this class? can be something simple explaining that it's a substitute for torch.*

dzhulgakov commented on torch/nn/quantized/modules/functional_modules.py:
```diff
@@ -0,0 +1,76 @@
+}
+
+class FloatFunctional(torch.nn.Module):
+    def __init__(self, op_name):
```
why do you need to initialize with op_name? I think it's much nicer API if you don't have to specify it - then your __init__ would be very simple. You could add some handling inside to ensure that only one method is ever called, but I'd say even that is optional.

raghuramank100 replied:
I had suggested this to protect from users doing
```
 # in init
self. f1 = nn.quantize.functionalwrapper()

def forward(self ,x, y):
   self.f1.add(x,y)
   self.f1.cat([x,x])
```
This should be an error: We cannot have f1 be used for multiple operations, as the observers have to be per instance.

Also, explicitly specifying the operation at init is consistent with how modules are used. A module does a fixed operation, defined at init. With the wrapper, it is hard to determine what the operations are looking at the init.

zafartahirov replied:
I am not sure what to do here -- we had a long discussion about that, so we need to make a decision soon if we need that or not.

dzhulgakov replied:
I'd vote for not specifying it in constructor and @soumith was voting for the same. It's strictly more lines of code and still not a complete protection as I can mix up two different adds together. The benefit of not specifying it in constructor is that one can build a nice wrapper like (for eager only, but still cute):

```
def __init__(...):
  self.stats = [nn.quantize.functionalwrapper() for i in range(10)]

def forward():
  counter = Counter()
  z = self.stats[counter.next()].add(x,y)
  return self.stats[counter.next()].cat([z,z])
```

We shouldn't provide it as default, but it feels more flexible.

We could remember the first op being invoked in a field on python side and re-check it next time. But it'd be interfere with torchscript a bit today.

raghuramank100 replied:
Agree with the comment about incomplete protection, but isn't some protection better than none?.  But do not understand the benefit of the wrapper. One could always do:
```
def __init__(...):
  opList = ['add', 'cat', ....]
  self.stats = [nn.quantize.functionalwrapper(opList[i]) for i in range(len(opList)]

def forward():
  counter = Counter()
  z = self.stats[counter.next()].add(x,y)
  return self.stats[counter.next()].cat([z,z])
```
@dzhulgakov : If you are convinced about having no argument in the constructor, lets go with it.

zafartahirov replied:
I am ok with either implementation. However, to clarify the purpose of the name in the constructor is to make sure that the functionalwrapper is used as intended in the constructor. I can see benefits of it from usage safety. But at the same time, I think this is getting into domain of over-protecting the user.

What I propose is, why don't we submit one or another, and discuss it in more details?

dzhulgakov replied:
I'll use @soumith as a tie-breaker - he voted for no name when we spoke earlier.

@raghuramank100 - in your example, it's still needed to align function types in two places (for example if one comments out some code)

raghuramank100 replied:
Yes, we still need to align function names in two places.  Lets go with removing the name as an argument. A todo will be to have torchscript ensure that observer instantiations are unique.

Commit:
Update on "Removing the make_module script."


n128675

Differential Revision: [D16593364](https://our.internmc.facebook.com/intern/diff/D16593364)

Commit:
Update on "Removing the make_module script."


n128675

Differential Revision: [D16593364](https://our.internmc.facebook.com/intern/diff/D16593364)

Commit:
Update on "Removing the make_module script."


n128675

Differential Revision: [D16593364](https://our.internmc.facebook.com/intern/diff/D16593364)

Commit:
Update on "Removing the make_module script."


n128675

Differential Revision: [D16593364](https://our.internmc.facebook.com/intern/diff/D16593364)
